File: src/main/java/com/iobestgroup/donkeymoney/user/UserService.java
package com.iobestgroup.donkeymoney.user;

import com.google.common.collect.Streams;
import com.iobestgroup.donkeymoney.user.exceptions.UserAlreadyExistsException;
import com.iobestgroup.donkeymoney.user.exceptions.UserDoesNotExistException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.stream.Collectors;


@Service
public class UserService {

    private final UserRepository userDao;

    @Autowired
    public UserService(UserRepository repository) {
        this.userDao = repository;
    }

    /**
     * @param potentialUser
     * @return potentialUser with id.
     * @throws UserAlreadyExistsException
     */
    public DMUser save(DMUser potentialUser) throws UserAlreadyExistsException {
        try {
            return userDao.save(potentialUser);
        } catch (DataIntegrityViolationException e){
            throw new UserAlreadyExistsException();
        }
    }

    public DMUser findByEmail(String email){
        DMUser user = userDao.findByEmail(email);
        if (user == null) throw new UserDoesNotExistException();
        return user;
    }

    /**
     * @return all users without theirs emails
     */
    public Iterable<DMUser> findAll(){
        return Streams.stream(userDao.findAll())
                .peek(it -> it.setEmail(null))
                .collect(Collectors.toList());
    }

    /**
     * @param search - must have at least 3 characters
     * @return empty list if search has less than 3 characters
     */
    public Iterable<DMUser> search(String search) {
        if (search.length() < 3) return new ArrayList<>();
        return userDao.search(search.toLowerCase() + '%');
    }
}

File: src/main/java/com/iobestgroup/donkeymoney/user/UserService.kt
package com.iobestgroup.donkeymoney.user

import com.iobestgroup.donkeymoney.user.exceptions.UserAlreadyExistsException
import com.iobestgroup.donkeymoney.user.exceptions.UserDoesNotExistException
import com.iobestgroup.donkeymoney.user.exceptions.UserNotAuthorizedException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service

@Service
class UserService @Autowired constructor(private val userDao: UserRepository) {

    /**
     * @param potentialUser
     * @return potentialUser with id.
     * @throws UserAlreadyExistsException
     */
    fun save(potentialUser: DMUser): DMUser = try {
//            potentialUser.password = BCrypt.hashpw(potentialUser.password, BCrypt.gensalt())
            userDao.save(potentialUser)
        } catch (e: DataIntegrityViolationException) {
            throw UserAlreadyExistsException()
    }


    fun findByEmail(email: String) = userDao.findByEmail(email)

    fun getSecurityToken(userName: String, password: String) = try {
        """
        |{
        |   "securityToken": "${userDao.getSecurityToken(userName, password).first().securityToken}"
        |}
        """.trimMargin()
    } catch (_ : NoSuchElementException){
        throw UserNotAuthorizedException()
    }


    fun findAll(): Iterable<DMUser> = userDao.findAll()


    fun update(user: DMUser) {
        try {
            val userToBeUpdated = userDao.findByEmail(user.email)
            userToBeUpdated.securityToken = user.securityToken
            userToBeUpdated.email = user.email
            userToBeUpdated.name = user.name
            userToBeUpdated.lastName = user.lastName
            userDao.save(userToBeUpdated)
        } catch (_: Exception){
            throw UserDoesNotExistException()
        }
    }

    fun deleteAll() {
        userDao.deleteAll()
    }
}


